### PR TITLE
Update ndarray schema to 1.1 to fix issue 345

### LINKED ISSUE #345: Overly permissive ndarray schema leads to improper time validation
The ndarray schema accepts any object that does not have a source key (and some that do). For example:
```python
import asdf
import asdf.tests


bad_string = """
!core/ndarray-1.0.0
 foo: 1
 bar: 2
"""

buff = asdf.tests.helpers.yaml_to_asdf(f"example: {bad_string.strip()}")

# confirm that validate on read is True
c = asdf.get_config()
c.validate_on_read = True

with asdf.AsdfFile() as af:
    try:
        af._open_impl(af, buff, mode='rw')
    except asdf.ValidationError as e:
        print("File appropriately failed to validate")
```
Does not produce a validation error and instead produces an exception during conversion to an array:
```
  File "/Users/bgraham/projects/221108_stdatamodel_bugs/asdf/asdf/tags/core/ndarray.py", line 395, in from_tree
    byteorder = node["byteorder"]
KeyError: 'byteorder
```

This schema is referenced in the time schema as part of an anyOf composition:
https://github.com/asdf-format/asdf-standard/blob/master/resources/schemas/stsci.edu/asdf/time/time-1.1.0.yaml#L127
which means that the subsequent object schema defined in time is not used and invalid yaml is produced for time objects that violate the time schema. For example, serializing
```python
astropy.time.Time(730120.0003703703, format="plot_date")
```
produces

```yaml
t: !time/time-1.1.0 {base_format: plot_date, value: '2000-01-01T00:00:32.000'}
```
which contains a base_format that is [not listed in the time schema](https://github.com/asdf-format/asdf-standard/blob/master/resources/schemas/stsci.edu/asdf/time/time-1.2.0.yaml#L180) (this example was taken from the time [tests in asdf-astropy](https://github.com/astropy/asdf-astropy/blob/main/asdf_astropy/converters/time/tests/test_time.py#L37) some of which are passing because of this bug).

### PR DESCRIPTION
**Pull request by braingram**
Fixes #345 

~~This PR should be merged after #349 so the time schema errors fixed in that PR do not result in test failures once this PR fixes #345.~~

### CODE REVIEW

**Comment by WilliamJamieson** [comment]
Can you separate the change adding the new `ndarray` schema into a separate commit followed by all the supporting changes to make the history a bit easier to parse.

Also, can you post the diff between `ndarray-1.0.0` and `ndarray-1.1.0`?

File: resources/schemas/stsci.edu/asdf/time/time-1.2.0.yaml [hunk 1]
```diff
@@ -178,24 +178,38 @@ definitions:
...
+      - ymdhms
+      - datetime64
 
 anyOf:
   - $ref: "#/definitions/string_formats"
 
   - $ref: "#/definitions/array_of_strings"
 
-  - $ref: "../core/ndarray-1.0.0#/anyOf/1"
-
```
Comment (WilliamJamieson) [inline resources/schemas/stsci.edu/asdf/time/time-1.2.0.yaml:198]: Isn't this PR supposed to make it so that the removal of this becomes unnecessary?

**Comment by braingram** [comment]
I may be missing something but I am unable to (locally) get the asdf-development tests to pass without updating the legacy converters to support the new schemas. Specifically:
- in astropy.io.misc.asdf)
  - table
  - column
- in asdf
  - integer
  - ndarray
 
 The failing test is `test_version_map_core_support[version6]` which fails with a rather unhelpful error:
`AttributeError: 'MockContext' object has no attribute '_warn_tag_mismatch'`
The MockContext [here](https://github.com/asdf-format/asdf/blob/master/asdf/tests/test_versioning.py#L279), meant to mimic an AsdfFile does not have that method. If a method is added (that does nothing, matching the default behavior in AsdfFile), the error becomes a more helpful:
```
E           AssertionError: ASDF Standard version 1.6.0 requires support for tag:stsci.edu:asdf/core/integer-1.1.0, but no ExtensionType exists that explicitly supports that version.
E           assert 'tag:stsci.ed...integer-1.0.0' == 'tag:stsci.ed...integer-1.1.0'
E             - tag:stsci.edu:asdf/core/integer-1.1.0
E             ?                                   ^
E             + tag:stsci.edu:asdf/core/integer-1.0.0
E             ?                                   ^
```
If I then add a `supported_versions` class attribute to the [IntegerType in asdf](https://github.com/asdf-format/asdf/blob/master/asdf/tags/core/integer.py#L46) the test no longer fails for the integer tag.

However, the NDArrayType class does not appear to support a similar fix. If I add a supported_versions class attribute I get an even less helpful error at time of class definition:
```python
TypeError: __class__ set to <class 'asdf.tags.core.ndarray.NDArrayType'> defining 'NDArrayType' as <class 'asdf.tags.core.ndarray.NDArrayType'>
```

To move this PR forward, am I correct that the asdf-development tests will fail until the converters are updated or am I missing something?

Also, thanks for the feedback on the history changes, I'll make them once the content of the PR is sorted out.

Aside from the examples in the description the diff for ndarray-1.0.0 and ndarray-1.1.0 is as follows:
```diff
>     oneOf:
>       - required: [source]
>       - required: [data]
>
```

**Comment by WilliamJamieson** [comment]
> I may be missing something but I am unable to (locally) get the asdf-development tests to pass without updating the legacy converters to support the new schemas. Specifically:
> 
> * in astropy.io.misc.asdf)
>   
>   * table
>   * column
> * in asdf
>   
>   * integer
>   * ndarray
> 
> The failing test is `test_version_map_core_support[version6]` which fails with a rather unhelpful error: `AttributeError: 'MockContext' object has no attribute '_warn_tag_mismatch'` The MockContext [here](https://github.com/asdf-format/asdf/blob/master/asdf/tests/test_versioning.py?rgh-link-date=2022-11-21T20%3A18%3A26Z#L279), meant to mimic an AsdfFile does not have that method. If a method is added (that does nothing, matching the default behavior in AsdfFile), the error becomes a more helpful:
> 
> ```
> E           AssertionError: ASDF Standard version 1.6.0 requires support for tag:stsci.edu:asdf/core/integer-1.1.0, but no ExtensionType exists that explicitly supports that version.
> E           assert 'tag:stsci.ed...integer-1.0.0' == 'tag:stsci.ed...integer-1.1.0'
> E             - tag:stsci.edu:asdf/core/integer-1.1.0
> E             ?                                   ^
> E             + tag:stsci.edu:asdf/core/integer-1.0.0
> E             ?                                   ^
> ```
> 
> If I then add a `supported_versions` class attribute to the [IntegerType in asdf](https://github.com/asdf-format/asdf/blob/master/asdf/tags/core/integer.py?rgh-link-date=2022-11-21T20%3A18%3A26Z#L46) the test no longer fails for the integer tag.
> 
> However, the NDArrayType class does not appear to support a similar fix. If I add a supported_versions class attribute I get an even less helpful error at time of class definition:
> 
> ```python
> TypeError: __class__ set to <class 'asdf.tags.core.ndarray.NDArrayType'> defining 'NDArrayType' as <class 'asdf.tags.core.ndarray.NDArrayType'>
> ```
> 
> To move this PR forward, am I correct that the asdf-development tests will fail until the converters are updated or am I missing something?
> 

This should be paired with an `asdf` PR. Failures in `astropy.io.misc.asdf` are irrelevant.

**Comment by braingram** [comment]
Here is a partial diff between ndarray-1.0.0 and ndarray-1.1.0 (most differences are updating references/tags from 1.0 to 1.1):
```diff
374a386,389
>     oneOf:
>       - required: [source]
>       - required: [data]
>
```

**Comment by braingram** [comment]
There is 1 remaining reference to ndarray-1.0 in the asdf-schema-1.0.0 metaschema:
https://github.com/asdf-format/asdf-standard/blob/1e223562401c64ca1b180356779c2082f0195310/resources/schemas/stsci.edu/asdf/asdf-schema-1.0.0.yaml#L39
This references a portion of the ndarray-1.0.0 schema that is not affected by #345.

The CI failure of asdf-development tests is unavoidable as asdf does not yet have support for the new schemas (support will be added in this PR: https://github.com/asdf-format/asdf/pull/1250).

I think it makes sense to merge this prior to the asdf PR so that the CI for that PR can hopefully have fewer errors (it will still fail some tests unless https://github.com/asdf-format/asdf/pull/1247 is merged to allow xfailing certain new tags that are not supported in asdf).

**Review by WilliamJamieson — APPROVED** [review]

**Pull request merged**